torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -128,31 +128,68 @@ at::Tensor pinnedLike(at::Tensor& tensor) {
 }
 
 // This function initializes a vector of CUDA streams, one for every
-// tensor in the input vector, and ensures that these streams are
+// tensor in the input tensor vector, and ensures that these streams are
 // synchronized with the current default streams. This is needed so
 // that new work on the new streams is serialized w.r.t. all operations
-// on the tensors in the input vector.
+// on the tensors.
 void initializeStreamsEvents(
-    std::vector<at::Tensor>& inputs,
+    std::vector<at::Tensor>& tensors,
     std::vector<at::cuda::CUDAStream>& streams,
     std::vector<at::cuda::CUDAEvent>& events) {
   at::cuda::OptionalCUDAGuard guard;
-  streams.reserve(inputs.size());
-  events.resize(inputs.size());
-  for (size_t i = 0; i < inputs.size(); i++) {
-    guard.set_index(inputs[i].get_device());
+  streams.reserve(tensors.size());
+  events.resize(tensors.size());
+  for (size_t i = 0; i < tensors.size(); i++) {
+    guard.set_index(tensors[i].get_device());
     // Record event on current stream
     events[i].record(at::cuda::getCurrentCUDAStream());
     // Get a non-default stream to execute asynchronous CUDA operations
-    // on for this input. This ensures that the default stream used
+    // on for this device. This ensures that the default stream used
     // by the caller is not occupied by c10d related operations.
     streams.push_back(at::cuda::getStreamFromPool(
-        /* isHighPriority */ true, inputs[i].get_device()));
+        /* isHighPriority */ true, tensors[i].get_device()));
     // Ensure the new stream is synchronized with the current stream.
     events[i].block(streams[i]);
   }
 }
 
+void initializeStreamsEvents(
+    std::vector<std::vector<at::Tensor>>& tensors,
+    std::vector<std::vector<int64_t>>& tensorStreamIndices,
+    std::vector<at::cuda::CUDAStream>& streams,
+    std::vector<at::cuda::CUDAEvent>& events) {
+  tensorStreamIndices.resize(tensors.size());
+  std::map<int64_t, size_t> streamMap;
+  for (size_t i = 0; i < tensors.size(); i++) {
+    tensorStreamIndices[i].resize(tensors[i].size());
+    for (size_t j = 0; j < tensors[i].size(); j++) {
+      auto deviceId = tensors[i][j].get_device();
+      if (streamMap.find(deviceId) == streamMap.end()) {
+        streamMap[deviceId] = streamMap.size();
+      }
+      tensorStreamIndices[i][j] = streamMap[deviceId];
+    }
+  }
+
+  streams.reserve(streamMap.size());
+  events.resize(streamMap.size());
+  at::cuda::OptionalCUDAGuard guard;
+  for (std::map<int64_t, size_t>::iterator it = streamMap.begin();
+       it != streamMap.end();
+       it++) {
```

Comment:
You can use `for (const auto& it : streamMap)` for short (no need to change, just FYI)

Reply:
updated